### PR TITLE
Update Envoy to efd9bef (Mar 19, 2024)

### PR DESCRIPTION
**Pull request by jiajunye**
- Update the `ENVOY_COMMIT` and `ENVOY_SHA` in `bazel/repositories.bzl` to the latest Envoy's commit.
- Update `source/client/process_impl.cc` to accommodate the new virtual functions created in https://github.com/envoyproxy/envoy/commit/f4b2d4a412dd70ea5afa74e81e1e320b5f2f5d88.
- Update `source/client/process_impl.cc` to accommodate the change of the constructor of `source/common/tls/context_manager_impl.h` in https://github.com/envoyproxy/envoy/commit/b7818b0df716af47ec22982c5a1cbbace5f2ae15.
- Update all the places that instantiate a `LoaderImpl` to accommodate the constructor change in https://github.com/envoyproxy/envoy/commit/708fa7b4d8269372fdac39b11caf2a3bf7b18d53.

### CODE REVIEW

**Comment by mum4k** [comment]
@fei-deng please review and assign to me once done.

**Review by fei-deng — APPROVED** [review]

**Review by mum4k — CHANGES_REQUESTED** [review]

File: source/client/process_impl.cc [hunk 1]
```diff
@@ -815,19 +822,20 @@ bool ProcessImpl::runInternal(OutputCollector& collector, const UriPtr& tracing_
...
     tls_.registerThread(*dispatcher_, true);
     store_root_.initializeThreading(*dispatcher_, tls_);
+    absl::Status creation_status;
     runtime_loader_ = Envoy::Runtime::LoaderPtr{new Envoy::Runtime::LoaderImpl(
         *dispatcher_, tls_, {}, *local_info_, store_root_, generator_,
-        Envoy::ProtobufMessage::getStrictValidationVisitor(), *api_)};
-    ssl_context_manager_ =
-        std::make_unique<Envoy::Extensions::TransportSockets::Tls::ContextManagerImpl>(
-            time_system_);
+        Envoy::ProtobufMessage::getStrictValidationVisitor(), *api_, creation_status)};
```
[the remark this answers is not in the document]
Comment (jiajunye) [reply source/client/process_impl.cc:828]: PTAL

File: source/client/process_impl.cc [hunk 2]
```diff
@@ -815,19 +822,25 @@ bool ProcessImpl::runInternal(OutputCollector& collector, const UriPtr& tracing_
...
+      return false;
+    }
 
     server_ = std::make_unique<NighthawkServerInstance>(
         admin_, *api_, *dispatcher_, access_log_manager_, envoy_options_, *runtime_loader_.get(),
         *singleton_manager_, tls_, *local_info_, validation_context_, grpc_context_,
         router_context_);
     server_factory_context_ =
         std::make_unique<NighthawkServerFactoryContext>(*server_, scope_root_);
+    ssl_context_manager_ =
```
Comment (mum4k) [inline source/client/process_impl.cc:841]: The `ssl_context_manager_` used to be initialized before the initialization of `server_`. This PR seems to be changing the order. Is the change necessary?

**Review by mum4k — COMMENTED** [review]
One more outstanding question. The rest looks good to me.

File: source/client/process_impl.cc (hunk 2 above)
Comment (jiajunye) [reply source/client/process_impl.cc:841]: Yes, I intentionally changed the order. `ssl_context_manager_` now requires `server_factory_context_` when instantiating. And `server_factory_context_` requires `server_` when instantiating.

Comment (mum4k) [reply source/client/process_impl.cc:841]: Thank you for explaining.

**Review by mum4k — APPROVED** [review]

**Pull request merged**